openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -1766,8 +1848,19 @@ static int cert_requested(SSL* ssl, X509** x509Out, EVP_PKEY** pkeyOut) {
 
     tcn_get_java_env(&e);
 
-    types = keyTypes(e, ssl);
-    issuers = principalBytes(e,  SSL_get_client_CA_list(ssl));
+ if (c->mode == SSL_MODE_SERVER) {
+        // TODO: Consider filling these somehow.
```

Comment:
Client's don't generally send servers a CA list in TLS. The `SSL_get_client_CA_list` function, as a server, would return back the configured list, rather than the list from the peer, which is a fairly different operation from what you want here.

TLS 1.3 does generalize the mechanism to work in both directions, but I'm not aware of anyone actually using it. (We don't process the extension in BoringSSL.) Usually the client has no end of supported CAs and the server already knows which to use for a given SNI, so there's not much negotiation to be done.

Moreover, although it's called `cert_requested`, this callback is used with `SSL_CTX_set_client_cert_cb`, which:
a) Is only used on a client anyway
b) Isn't what you want for `TLS_with_buffers_method` since it depends on `X509`.

I'm guessing this is why you had to build with `NDEBUG`? It fires the assert because you're not supposed to use it here. :-P You won't get the binary size benefits and, in some cases (though I think not this one right now), might end up leaking because `SSL_CTX`, etc., objects made from a non-`X509` `SSL_METHOD` don't know how to release `X509` objects hanging off of them.

You probably want `SSL_CTX_set_cert_cb` which works more uniformly (client/server, no X509 dependency, lets you set a cert chain, etc). (We actually emulate `SSL_CTX_set_client_cert_cb` using `SSL_CTX_set_cert_cb` anyway.)

Reply:
@davidben this function is never used with BoringSSL, it is just used when compiling against openssl < 1.0.2. 

I did compile with `NDEBUG` to reduce some overhead, it also works without specify it. Do you say I should remove this option ?

Reply:
sorry.. I wanted to say we never call `setCertRequestedCallback` these days as we replaced it with `SSL_CTX_set_cert_cb `. Let me just remove this all together.

Reply:
> I did compile with `NDEBUG` to reduce some overhead, it also works without specify it. Do you say I should remove this option?

I guess it's `assert`, so you probably want to build with it in release mode and without it in debug mode.